<a href="https://colab.research.google.com/github/nshamid/Metode-GADA-Grey-Absolute-Decision-Analysis-/blob/main/GADA_KELOMPOK5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## SISTEM PENDUKUNG KEPUTUSAN - KELOMPOK 5
### Metode GADA (Grey Absolute Decision Analysis)

### Studi Kasus


> Banyaknya pilihan game MOBA Mobile dengan karakteristik dan fitur yang beragam membuat mahasiswa sering kesulitan menentukan game mana yang paling sesuai untuk dimainkan bersama teman.





### Data Kriteria

| Kode | Kriteria | Jenis |
| :---: | :--- | :---: |
| C1 | Grafis & desain | Benefit |
| C2 | Keseimbangan gameplay | Benefit |
| C3 | Popularitas & komunitas | Benefit |
| C4 | Stabilitas server/koneksi | Benefit |
| C5 | Fitur & mode | Benefit |
| C6 | Kemudahan kontrol & UI | Benefit |
| C7 | Tingkat kompetitif | Benefit |
| C8 | Sistem pembayaran | Benefit |
| C9 | Ukuran aplikasi | Benefit |
| C10 | Lancar di HP menengah | Benefit |
| C11 | Fairness F2P | Benefit |
| C12 | Sosialisasi / menambah kenalan | Benefit |
| C13 | Rekomendasi untuk pemula | Benefit |

### Data Alternatif

| Kode | Alternatif |
| :---: | :--- |
| A1 | Mobile Legends: Bang Bang |
| A2 | Honor of Kings |
| A3 | Garena AOV |
| A4 | League of Legends: Wild Rift |
| A5 | Pokémon Unite |
| A6 | Heroes Evolved |
| A7 | Onmyoji Arena |
| A8 | Lokapala |
| A9 | Vainglory (IOS) |
| A10 | Marvel Super War |

### Import Libraries

In [ ]:
import numpy as np
import pandas as pd

### Tahap 1: Menghitung Nilai Geometrik Mean

In [ ]:
data_awal = {
    'Alternatif': ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'A10'],
    'C1': [4.207949381, 4.641588834, 3.55655882, 4.472135955, 2.828427125, 4, 4, 3, 2, 4],
    'C2': [3.78329541, 3.914867641, 3.55655882, 3.464101615, 2.449489743, 4, 4, 3, 2, 3.634241186],
    'C3': [4.295307446, 3.301927249, 2.91295063, 4, 2.449489743, 2, 3, 3, 3, 3.301927249],
    'C4': [3.590914704, 3.107232506, 2.91295063, 3.872983346, 2.828427125, 4, 4, 3, 3, 3.634241186],
    'C5': [4.278081229, 4.641588834, 3.13016816, 4.472135955, 2.449489743, 4, 4, 4, 3, 3.634241186],
    'C6': [4.344938193, 4.641588834, 2.91295063, 4.472135955, 2.828427125, 3, 4, 4, 2, 3.634241186],
    'C7': [4.257678109, 3.914867641, 3.55655882, 4.472135955, 2, 4, 4, 4, 2, 3.634241186],
    'C8': [4.065840375, 4.641588834, 3.55655882, 3, 2.828427125, 4, 4, 4, 3, 3.634241186],
    'C9': [4.0622705, 2.884999141, 2.632148026, 3.872983346, 2.828427125, 4, 4, 4, 3, 3.634241186],
    'C10': [3.650022501, 4.844999141, 2.91295063, 2.828427125, 2.449489743, 3, 4, 4, 3, 3.634241186],
    'C11': [3.799034928, 4.641588834, 3.080070098, 3.872983346, 2.449489743, 4, 4, 4, 3, 3.634241186],
    'C12': [3.702746903, 2.620741394, 1.861209718, 1.44213562, 2.449489743, 2, 3, 3, 3, 3.634241186],
    'C13': [3.456730406, 2.620741394, 2.51486659, 1.732050808, 2.828427125, 3, 3, 2, 3, 4]
}
df_awal = pd.DataFrame(data_awal)

# Pisahkan kolom Alternatif dan Kriteria
df_alternatif = df_awal[['Alternatif']]
df_kriteria = df_awal.drop(columns=['Alternatif'])

In [ ]:
gm_max = df_kriteria.max()

df_gm_max = pd.DataFrame(gm_max, columns=['GM Max'])
display(df_gm_max)

GM Max
C1  4.641588834
C2  4.000000000
C3  4.295307446
C4  4.000000000
C5  4.641588834
C6  4.641588834
C7  4.472135955
C8  4.641588834
C9  4.062270500
C10 4.844999141
C11 4.641588834
C12 3.702746903
C13 4.000000000

### Tahap 2: Menentukan Nilai Matriks Perbandingan Berpasangan



> Langkah 2.1 Menghitung Nilai Komponen Alternatif



In [ ]:
df_normalisasi = df_kriteria.div(gm_max, axis=1)

df_hasil_normalisasi = pd.concat([df_alternatif, df_normalisasi], axis=1)
display(df_hasil_normalisasi)

Alternatif          C1          C2          C3          C4          C5  \
0         A1 0.906575212 0.945823853 1.000000000 0.897728676 0.921684661   
1         A2 1.000000000 0.978716910 0.768728965 0.776808126 1.000000000   
2         A3 0.766237370 0.889139705 0.678170461 0.728237658 0.674374287   
3         A4 0.963492484 0.866025404 0.931248822 0.968245836 0.963492484   
4         A5 0.609366152 0.612372436 0.570271110 0.707106781 0.527726567   
5         A6 0.861773876 1.000000000 0.465624411 1.000000000 0.861773876   
6         A7 0.861773876 1.000000000 0.698436617 1.000000000 0.861773876   
7         A8 0.646330407 0.750000000 0.698436617 0.750000000 0.861773876   
8         A9 0.430886938 0.500000000 0.698436617 0.750000000 0.646330407   
9        A10 0.861773876 0.908560297 0.768728965 0.908560297 0.782973528   

           C6          C7          C8          C9         C10         C11  \
0 0.936088557 0.952045768 0.875958755 1.000000000 0.753358751 0.818477264   
1 1.000000000 0.875391017 1.000000000 0.710193755 1.000000000 1.000000000   
2 0.627576189 0.795270729 0.766237370 0.647949964 0.601228307 0.663580987   
3 0.963492484 1.000000000 0.646330407 0.953403607 0.583782792 0.834408967   
4 0.609366152 0.447213595 0.609366152 0.696267549 0.505570728 0.527726567   
5 0.646330407 0.894427191 0.861773876 0.984671011 0.619195156 0.861773876   
6 0.861773876 0.894427191 0.861773876 0.984671011 0.825593542 0.861773876   
7 0.861773876 0.894427191 0.861773876 0.984671011 0.825593542 0.861773876   
8 0.430886938 0.447213595 0.646330407 0.738503258 0.619195156 0.646330407   
9 0.782973528 0.812641034 0.782973528 0.894632986 0.750101513 0.782973528   

          C12         C13  
0 1.000000000 0.864182602  
1 0.707783022 0.655185348  
2 0.502656478 0.628716648  
3 0.389477233 0.433012702  
4 0.661533129 0.707106781  
5 0.540139538 0.750000000  
6 0.810209306 0.750000000  
7 0.810209306 0.500000000  
8 0.810209306 0.750000000  
9 0.981498677 1.000000000



> Langkah 2.2 Menghitung Matriks Perbandingan Berpasangan



In [ ]:
# Ambil nilai normalisasi
df_normalized_values = df_hasil_normalisasi.drop(columns=['Alternatif']).copy()

# --- Hitung Kolom '1/2 x C1' ---
df_normalized_values['1/2 x C1'] = df_normalized_values['C1'] / 2

# --- Hitung Kolom 'SUM C2-C12' ---
df_normalized_values['SUM C2-C12'] = df_normalized_values['C2'] + df_normalized_values['C12']

# --- Hitung Kolom '1/2 x C13' ---
df_normalized_values['1/2 x C13'] = df_normalized_values['C13'] / 2

# --- Hitung Kolom 'Si' ---
df_normalized_values['Si'] = (
    df_normalized_values['1/2 x C1'] +
    df_normalized_values['SUM C2-C12'] +
    df_normalized_values['1/2 x C13']
)

df_tabel = df_normalized_values[['1/2 x C1', 'SUM C2-C12', '1/2 x C13', 'Si']].copy()
df_tabel_final = pd.concat([df_awal[['Alternatif']], df_tabel], axis=1)
display(df_tabel_final)

Alternatif    1/2 x C1  SUM C2-C12   1/2 x C13          Si
0         A1 0.453287606 1.945823853 0.432091301 2.831202759
1         A2 0.500000000 1.686499933 0.327592674 2.514092607
2         A3 0.383118685 1.391796183 0.314358324 2.089273192
3         A4 0.481746242 1.255502637 0.216506351 1.953755230
4         A5 0.304683076 1.273905564 0.353553391 1.932142031
5         A6 0.430886938 1.540139538 0.375000000 2.346026476
6         A7 0.430886938 1.810209306 0.375000000 2.616096244
7         A8 0.323165203 1.560209306 0.250000000 2.133374510
8         A9 0.215443469 1.310209306 0.375000000 1.900652775
9        A10 0.430886938 1.890058973 0.500000000 2.820945911



> Langkah 2.3 Menghitung Elemen Matriks Perbandingan Berpasangan



In [ ]:
data_si = {
    'Alternatif': ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8', 'A9', 'A10'],
    'Si': [2.833702759, 2.514092606, 2.089273192, 1.953755230, 1.932142032, 2.346026476, 2.616096244, 2.133374509, 1.900652775, 2.820945911]
}
df_si = pd.DataFrame(data_si)

In [ ]:
Si_values = df_si['Si'].values
alternatif_labels = df_si['Alternatif'].values
n = len(Si_values)

epsilon_matrix = np.zeros((n, n))

# --- Perhitungan Matriks Perbandingan Berpasangan (epsilon_ij) ---
for i in range(n):
    for j in range(n):
        Si = Si_values[i]  # Baris (i)
        Sj = Si_values[j]  # Kolom (j)

        # RUMUS SPESIFIK: (1 + Si + Sj) / (1 + Sj + (Si - Sj))

        # Pembilang (Numerator)
        numerator = 1 + Si + Sj

        # Penyebut (Denominator)
        denominator = 1 + Sj + (Si - Sj)

        # Hasil epsilon_ij
        epsilon_matrix[i, j] = numerator / denominator

# --- Konversi ke DataFrame dan Pembulatan ---
df_epsilon = pd.DataFrame(
    epsilon_matrix,
    index=alternatif_labels,  # Baris (i)
    columns=alternatif_labels # Kolom (j)
).round(9)

# Menambahkan nama untuk Index dan Columns
df_epsilon.index.name = 'Alternatif'
display(df_epsilon)

A1          A2          A3          A4          A5  \
Alternatif                                                               
A1         1.739155573 1.655787046 1.544975269 1.509626164 1.503988481   
A2         1.806382494 1.715431518 1.594541302 1.555977161 1.549826726   
A3         1.917271663 1.813813622 1.676299266 1.632432002 1.625435794   
A4         1.959355985 1.851151301 1.707327801 1.661447912 1.654130719   
A5         1.966427522 1.857425247 1.712541606 1.666323530 1.658952401   
A6         1.846885935 1.751366621 1.624404262 1.583903099 1.577443737   
A7         1.783635879 1.695250468 1.577770350 1.540294035 1.534317093   
A8         1.904361337 1.802359437 1.666780554 1.623530709 1.616632971   
A9         1.976918983 1.866733388 1.720276901 1.673557086 1.666105936   
A10        1.741623364 1.657976497 1.546794757 1.511327633 1.505671129   

                    A6          A7          A8          A9         A10  
Alternatif                                                              
A1         1.611947932 1.682394126 1.556478852 1.495774684 1.735828020  
A2         1.667605194 1.744458538 1.607091146 1.540865876 1.802752297  
A3         1.759410492 1.846832275 1.690574895 1.615242698 1.913142262  
A4         1.794252162 1.885684845 1.722258394 1.643469965 1.955037128  
A5         1.800106697 1.892213343 1.727582254 1.648213065 1.962076830  
A6         1.701138049 1.781851627 1.637584468 1.568032796 1.843073398  
A7         1.648773240 1.723458688 1.589966186 1.525609012 1.780108084  
A8         1.748722015 1.834913361 1.680855258 1.606583340 1.900290056  
A9         1.808792592 1.901899140 1.735480829 1.655250015 1.972521060  
A10        1.613991020 1.684672410 1.558336747 1.497429909 1.738284701

### Tahap 3: Menghitung Bobot Yang Disarankan

In [ ]:
# 1. Menghitung Hasil Kali (Perkalian tiap baris)
df_epsilon['Hasil Kali'] = df_epsilon.prod(axis=1)

# 2. Menghitung GM Baris (Rata-rata Geometrik)
df_epsilon['GM Baris'] = df_epsilon['Hasil Kali']**(1/n)

# 3. Menghitung Total (Sigma) GM Baris
total_gm_baris = df_epsilon['GM Baris'].sum()

# 4. Menghitung Bobot Alternatif (Normalisasi)
df_epsilon['Bobot Alternatif'] = df_epsilon['GM Baris'] / total_gm_baris

# Total Bobot Alternatif
total_bobot_alternatif = df_epsilon['Bobot Alternatif'].sum()

# Membuat DataFrame Hasil Akhir
df_hasil = df_epsilon[['Hasil Kali', 'GM Baris', 'Bobot Alternatif']].copy()
df_hasil.index.name = 'Alternatif'
df_hasil = df_hasil.reset_index()

# Membuat baris Total
total_row = pd.DataFrame({
    'Alternatif': ['Total'],
    'Hasil Kali': [''],
    'GM Baris': [total_gm_baris],
    'Bobot Alternatif': [total_bobot_alternatif]
})

df_final = pd.concat([df_hasil, total_row], ignore_index=True)

# Menampilkan Hasil Perhitungan
def format_cell(value):
    if isinstance(value, (float, np.float64)):
        if value == total_gm_baris:
            return f'{value:.9f}'
        elif value == total_bobot_alternatif:
            return f'{value:.0f}'
        elif 'GM Baris' in df_final.columns and value in df_final['GM Baris'].values:
            return f'{value:.9f}'
        elif 'Bobot Alternatif' in df_final.columns and value in df_final['Bobot Alternatif'].values:
            return f'{value:.12f}'
        elif 'Hasil Kali' in df_final.columns and value in df_final['Hasil Kali'].values:
             return f'{value:.8f}'
    return str(value)

df_final_formatted = df_final.applymap(format_cell)

display(df_final_formatted)

/tmp/ipython-input-1641846101.py:46: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_final_formatted = df_final.applymap(format_cell)


Alternatif    Hasil Kali      GM Baris Bobot Alternatif
0          A1  110.70652519   1.601095814   0.094044525686
1          A2  154.73926580   1.655617671   0.097247008761
2          A3  262.57107804   1.745519435   0.102527622619
3          A4  318.63715371   1.779629685   0.104531176850
4          A5  329.04865216   1.785360868   0.104867812789
5          A6  188.32428565   1.688459441   0.099176055525
6          A7  138.33814060   1.637171566   0.096163528828
7          A8  247.24694691   1.735054391   0.101912931048
8          A9  345.05950106   1.793863524   0.105367238355
9         A10  112.09838346   1.603097493   0.094162099541
10      Total                17.024869889                1



### Tahap 4: Menghitung Bobot Simulasi dan Kriteria



In [ ]:
n = 10
kriteria_labels = [f'C{i}' for i in range(1, n + 1)]
df_epsilon.index = kriteria_labels

df_epsilon_criteria = pd.DataFrame(
    epsilon_matrix,
    index=kriteria_labels,
    columns=kriteria_labels
).round(9)

# 1. Menghitung Hasil Kali (Perkalian tiap kolom)
df_epsilon_criteria['Hasil Kali'] = df_epsilon_criteria.prod(axis=0)

# 2. Menghitung GM Kolom (Rata-rata Geometrik)
df_epsilon_criteria['GM Kolom'] = df_epsilon_criteria['Hasil Kali']**(1/n)

# 3. Menghitung Total (Sigma) GM Kolom
total_gm_kolom = df_epsilon_criteria['GM Kolom'].sum()

# 4. Menghitung Bobot Kriteria (Normalisasi)
df_epsilon_criteria['Bobot kriteria'] = df_epsilon_criteria['GM Kolom'] / total_gm_kolom

# Total Bobot Kriteria
total_bobot_kriteria = df_epsilon_criteria['Bobot kriteria'].sum()

# Membuat DataFrame Hasil Akhir
df_hasil = df_epsilon_criteria[['Hasil Kali', 'GM Kolom', 'Bobot kriteria']].copy()
df_hasil.index.name = 'Kriteria'
df_hasil = df_hasil.reset_index()

# Membuat baris Total
total_row = pd.DataFrame({
    'Kriteria': ['Total'],
    'Hasil Kali': [''],
    'GM Kolom': [total_gm_kolom],
    'Bobot kriteria': [total_bobot_kriteria]
})

df_final = pd.concat([df_hasil, total_row], ignore_index=True)

# Menampilkan Hasil Perhitungan
def format_cell(value):
    if isinstance(value, (float, np.float64)):
        if value == total_gm_kolom:
            return f'{value:.9f}'
        elif value == total_bobot_kriteria:
            return f'{value:.0f}'
        elif 'GM Kolom' in df_final.columns and value in df_final['GM Kolom'].values:
            return f'{value:.9f}'
        elif 'Bobot kriteria' in df_final.columns and value in df_final['Bobot kriteria'].values:
            return f'{value:.12f}'
        elif 'Hasil Kali' in df_final.columns and value in df_final['Hasil Kali'].values:
             return f'{value:.8f}'
    return str(value)

df_final_formatted = df_final.applymap(format_cell)

display(df_final_formatted)

/tmp/ipython-input-2721881384.py:56: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_final_formatted = df_final.applymap(format_cell)


Kriteria    Hasil Kali      GM Kolom  Bobot kriteria
0        C1  501.25860039   1.862113650  0.109267673634
1        C2  293.38425105   1.764995814  0.103568859258
2        C3  137.25061151   1.635879947  0.095992420308
3        C4  106.35200781   1.594683752  0.093575053143
4        C5  102.05032577   1.588113142  0.093189493793
5        C6  218.72796011   1.713919541  0.100571735232
6        C7  349.17728681   1.795992830  0.105387744945
7        C8  148.92178761   1.649285436  0.096779046098
8        C9   96.06289630   1.578539914  0.092627742732
9       C10  490.92199675   1.858237625  0.109040230856
10    Total                17.041761651               1

### Tahap 5: Menghitung Agregasi Bobot Terhadap Kriteria

In [ ]:
df_alternatif_labels = df_awal[['Alternatif']].copy()

bobot_alternatif_values = [
    0.094044525686, 0.097247008761, 0.102527622619, 0.104531176850,
    0.104867812789, 0.099176055525, 0.096163528828, 0.101912931048,
    0.105367238355, 0.094162099541
]

# Total Bobot Alternatif (seharusnya 1.0)
total_bobot_alternatif = sum(bobot_alternatif_values)


# 1. Menggabungkan label Alternatif A1-A10 dengan Bobot yang sudah dihitung
df_gada = df_alternatif_labels.copy()
df_gada['Bobot Alternatif'] = bobot_alternatif_values

# 2. Membuat kolom Bobot GADA
df_gada['Bobot GADA'] = df_gada['Bobot Alternatif']

# 3. Membuat Baris Total
row_total_gada = pd.DataFrame({
    'Alternatif': ['Total'],
    'Bobot Alternatif': [total_bobot_alternatif],
    'Bobot GADA': [total_bobot_alternatif]
})

# 4. Menggabungkan data dengan baris total
df_final_gada = pd.concat([df_gada, row_total_gada], ignore_index=True)

# Fungsi format (Dipertahankan)
def format_gada(value):
    if isinstance(value, (float, np.float64)):
        # Jika nilainya 1 (Total), tampilkan angka bulat
        if round(value, 5) == 1.0:
            return f'{value:.0f}'
        # Selain itu tampilkan desimal panjang (misal 8 digit)
        return f'{value:.8f}'
    return str(value)

df_gada_formatted = df_final_gada.applymap(format_gada)

display(df_gada_formatted)

/tmp/ipython-input-1700551721.py:40: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_gada_formatted = df_final_gada.applymap(format_gada)


Alternatif Bobot Alternatif  Bobot GADA
0          A1       0.09404453  0.09404453
1          A2       0.09724701  0.09724701
2          A3       0.10252762  0.10252762
3          A4       0.10453118  0.10453118
4          A5       0.10486781  0.10486781
5          A6       0.09917606  0.09917606
6          A7       0.09616353  0.09616353
7          A8       0.10191293  0.10191293
8          A9       0.10536724  0.10536724
9         A10       0.09416210  0.09416210
10      Total                1           1

### Tahap 6: Hasil Perangkingan

| RANK | Kode | Alternatif | Bobot Akhir | Keterangan |
| :---: | :---: | :--- | :---: | :--- |
| 1 | **A9** | **Vainglory (IOS)** | **0.10536724** | **Pilihan Terbaik (Bobot Tertinggi)** |
| 2 | A5 | Pokémon Unite | 0.10486781 | |
| 3 | A4 | League of Legends: Wild Rift | 0.10453118 | |
| 4 | A3 | Garena AOV | 0.10252762 | |
| 5 | A8 | Lokapala | 0.10191293 | |
| 6 | A6 | Heroes Evolved | 	0.09917606 | |
| 7 | A2 | Honor of Kings | 0.09724701 | |
| 8 | A7 | Onmyoji Arena | 0.09616353 | |
| 9 | A10 | Marvel Super War | 0.09416210 | |
| 10 | **A1** | Mobile Legends: Bang Bang | **0.09404453** | **Pilihan Terburuk (Bobot Terendah)** |
